### PR TITLE
Add remaining bbox utils

### PR DESCRIPTION
**Pull request by sineeli**
Add remaining utils related to bounding box.

### CODE REVIEW

File: keras_nlp/src/bounding_box/mask_invalid_detections.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            "`bounding_box.to_dense(bounding_boxes)` before passing your boxes "
+            "to `bounding_box.mask_invalid_detections()`."
+        )
+    if "num_detections" not in bounding_boxes:
+        raise ValueError(
+            "`bounding_boxes` must have key 'num_detections' "
+            "to be used with `bounding_box.mask_invalid_detections()`."
+        )
+
+    boxes = bounding_boxes.get("boxes")
```
Comment (mattdangerw) [inline keras_nlp/src/bounding_box/mask_invalid_detections.py:63]: Where is this expected to come from? Is a dict with this structure standard for certain models? What generates this?

Probably we should add this along with the model code that needs it. And possibly unexposed.

**Review by mattdangerw — COMMENTED** [review]
Just a lot of question for now. I suspect we should probably just land the bbox stuff for now (which we won't release here anyway), skip the layers (or keep them in a model dir), and start working through an individual model.

File: keras_nlp/src/bounding_box/ensure_tensor.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.utils import tensor_utils
+
+
+@keras_nlp_export("keras_nlp.bounding_box.ensure_tensor")
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_nlp/src/bounding_box/ensure_tensor.py:19]: Yes, I thought so as well but this was handling dictionary format of inputs, we can add this at model level.

File: keras_nlp/src/layers/modeling/anchor_generator.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+import keras
+from keras import ops
+
+from keras_nlp.src.api_export import keras_nlp_export
+from keras_nlp.src.bounding_box import converters
+
+
+@keras_nlp_export("keras_nlp.layers.AnchorGenerator")
+class AnchorGenerator(keras.layers.Layer):
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_nlp/src/layers/modeling/anchor_generator.py:25]: `AnchorGenerator`, `NonMaxSupression` and `BoxMatcher `layers are used in all Object Detection Models.

* YOLO
* Faster R-CNN
* RetinaNet
* Mas R-CNN

If any extra Object Detection Models are added in future these are mandatory layers across all these models. So added as part of layers.

If these layers are available from core directly we can isolate these to only particular model dir `RetinaNet` for now.

File: keras_nlp/src/utils/tensor_utils.py [hunk 4]
```diff
@@ -204,3 +204,12 @@ def any_equal(inputs, values, padding_mask):
         output = ops.logical_or(output, value_equality)
 
     return ops.logical_and(output, padding_mask)
+
+
+def ensure_tensor(inputs, dtype=None):
```
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_nlp/src/utils/tensor_utils.py:209]: yes, just for dictionary scenario

File: keras_nlp/src/bounding_box/mask_invalid_detections.py (hunk 1 above)
Comment (sineeli) [reply keras_nlp/src/bounding_box/mask_invalid_detections.py:63]: Object Detection models has standard two inputs as dictionary format input and output which are basically `images` and `boudning_boxes` itself is another dictionary having `boxes` and `classes` as keys.

So for any model which return bounding box outputs this util is used to mask invalid outputs and  mark them as `-1`.

File: keras_nlp/src/layers/modeling/anchor_generator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_nlp/src/layers/modeling/anchor_generator.py]: SG! Well I think adding these in a separate PR will definitely be easier, and along with a model should help us suss out the task setup. I think we probably need so more substantial rewrites before we exposed these. In particular:

- I doubt we want to rely on raggeds as much. Let's write this the "jax and torch" way. So if we have ragged inputs, we probably want to convert them to padded dense inputs as a first step.
- We should use common test cases in the base test case, like our other layers.
- We should rewrite docstrings, etc to match local style.

The first of these is probably the trickiest, but it will be easiest to do when we have a actual model to work with.

**Review by mattdangerw — APPROVED** [review]
lgtm! not looking at the code too closely as I believe we are still going to lower this to keras, so this is just a temp unblocker

**Comment by sineeli** [comment]
> lgtm! not looking at the code too closely as I believe we are still going to lower this to keras, so this is just a temp unblocker

Sure

File: keras_nlp/src/layers/modeling/anchor_generator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sineeli) [reply keras_nlp/src/layers/modeling/anchor_generator.py]: @mattdangerw,

So in the next PR make changes as per "jax and torch" for layers and write these layers as of now at model directory.

I will try to write one layer at a time in the PR and follow the points who have mentioned.

Thanks!
